system/modules/core/library/Contao/Combiner.php

```diff
@@ -383,48 +383,43 @@ protected function fixPaths($content, $arrFile)
 		$strDirname = dirname($arrFile['name']);
 		$strGlue = ($strDirname != '.') ? $strDirname . '/' : '';
 
-		$strBuffer = '';
-		$chunks = preg_split('/url\((["\']?+)(.+?)\1\)/', $content, -1, PREG_SPLIT_DELIM_CAPTURE);
-
-		// Check the URLs
-		for ($i=0, $c=count($chunks); $i<$c; $i=$i+3)
+		return preg_replace_callback('/url\(("[^"\n]+"|\'[^\'\n]+\'|[^"\'\s()]+)\)/', function($matches) use ($strDirname, $strGlue)
 		{
-			$strBuffer .= $chunks[$i];
-
-			if (!isset($chunks[$i+2]))
+			$strData = $matches[1];
+			if ($strData[0] === '"' || $strData[0] === '\'')
 			{
-				break;
+				$strData = substr($strData, 1, -1);
 			}
 
-			$strData = $chunks[$i+2];
-
 			// Skip absolute links and embedded images (see #5082)
-			if (strncmp($strData, 'data:', 5) !== 0 && strncmp($strData, 'http://', 7) !== 0 && strncmp($strData, 'https://', 8) !== 0 && strncmp($strData, '/', 1) !== 0 && strncmp($strData, 'assets/css3pie/', 15) !== 0)
+			if (strncmp($strData, 'data:', 5) === 0 || strncmp($strData, 'http://', 7) === 0 || strncmp($strData, 'https://', 8) === 0 || strncmp($strData, '/', 1) === 0 || strncmp($strData, 'assets/css3pie/', 15) === 0)
 			{
-				// Make the paths relative to the root (see #4161)
-				if (strncmp($strData, '../', 3) !== 0)
-				{
-					$strData = '../../' . $strGlue . $strData;
-				}
-				else
-				{
-					$dir = $strDirname;
+				return $matches[0];
+			}
 
-					// Remove relative paths
-					while (strncmp($strData, '../', 3) === 0)
-					{
-						$dir = dirname($dir);
-						$strData = substr($strData, 3);
-					}
+			// Make the paths relative to the root (see #4161)
+			if (strncmp($strData, '../', 3) !== 0)
+			{
+				$strData = '../../' . $strGlue . $strData;
+			}
+			else
+			{
+				$dir = $strDirname;
 
-					$glue = ($dir != '.') ? $dir . '/' : '';
-					$strData = '../../' . $glue . $strData;
+				// Remove relative paths
+				while (strncmp($strData, '../', 3) === 0)
+				{
+					$dir = dirname($dir);
+					$strData = substr($strData, 3);
 				}
+
+				$glue = ($dir != '.') ? $dir . '/' : '';
+				$strData = '../../' . $glue . $strData;
 			}
 
-			$strBuffer .= 'url(' . $chunks[$i+1] . $strData . $chunks[$i+1] . ')';
-		}
+			$strQuote = $matches[1][0] === '\'' ? '\'' : '"';
```

Comment:
This would always add double quotes even if the original string had no quotes at all.

Reply:
Yes, that’s on purpose. We change the path so we should also add quotes (as we did in the past).

Reply:
Did we really do this in the past? Because we really should not! Two unnecessary characters that have not been there before.

Reply:
But they may be necessary, depending on the path we generate.

Yes, in the past we always used double quotes: https://github.com/contao/core/blob/3.5.28/system/modules/core/library/Contao/Combiner.php#L425

Reply:
Yes, but that does no longer work with unencoded `data:` URLs, which is why I have changed it  (see #8788). We must use single quotes if the original used single quotes, double quotes if the original used double quotes and we can and should use no quotes if the original used no quotes.

Reply:
If `data:` is used, the replacement gets completley skipped and nothing gets changed at all (in line 397).

If we replace `url(file.jpg)` with e.g. `url(../../files/some dir/file.jpg)` we would break it because whitespace (and some other characters) are not allowed in unquoted URLs.

Reply:
Then we should probably check this? I don't know, its two unnecessary characters. 😄

Reply:
I think it’s unnecessary to remove the quotes. For the gzipped size it could even be better if all `url()`s use quotes :)

Reply:
This would do the trick:

```php
			if ($matches[1][0] != '"' && $matches[1][0] != "'" && !preg_match('/[()"\' ]/', $strDirname) && !preg_match('/[()"\' ]/', $strData))
			{
				$strQuote = '';
			}
			else
			{
				$strQuote = $matches[1][0] == "'" ? "'" : '"';
			}
```